File: yangyiyang/average_of_levels_in_binary_tree.cpp
 /**
  *  * Definition for a binary tree node.
  *   * struct TreeNode {
  *    *     int val;
  *     *     TreeNode *left;
  *      *     TreeNode *right;
  *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  *        * };
  *         */
class Solution {
    public:
    vector<double> averageOfLevels(TreeNode* root) {
            vector<double> result;

            queue<TreeNode*> que;
            que.push(root);
            TreeNode *temp;

            while(!que.empty())
            {
                        int count = que.size();
                        double sum = 0;

                        for(int i = 0; i < count; i++)
                        {
                                        temp = que.front();
                                        que.pop();

                                        sum += temp->val;
                                        if(temp->left != NULL) que.push(temp->left);
                                        if(temp->right != NULL) que.push(temp->right);
                                    }

                        result.push_back(sum/count);
                    }

            return result;
        }

};

File: caoguodong/Average of Levels in Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        List<Double> resultList = new ArrayList<>();
        queue.offer(root);
        queue.offer(null);
        double sum = 0;
        int size = 0;
        while(queue.size()>0){
            TreeNode treeNode = queue.poll();
            if(treeNode != null){
                if(treeNode.left != null){
                    queue.offer(treeNode.left);   
                }
                
                if(treeNode.right != null){
                    queue.offer(treeNode.right);
                }
                sum = sum + treeNode.val;
                size++;
            } else{
                resultList.add(sum/size);
                sum = 0;
                size =0;
                if(queue.size() > 0){
                    queue.offer(null);   
                }
            }
        }
        return resultList;
    }
}